agent/stream.go

```diff
@@ -0,0 +1,129 @@
+package agent
+
+import (
+	"sync"
+
+	"github.com/odpf/meteor/models"
+	"github.com/pkg/errors"
+)
+
+type streamMiddleware func(src models.Record) (dst models.Record, err error)
+type subscriber struct {
+	callback  func([]models.Record) error
+	channel   chan models.Record
+	batchSize int
+}
+
+type stream struct {
+	middlewares []streamMiddleware
+	subscribers []*subscriber
+	hasClosed   bool
+	err         error
+}
+
+func newStream() *stream {
+	return &stream{}
+}
+
+// subscribe() will register callback with a batch size to the emitter.
+// Calling this will not start listening yet, use broadcast() to start sending data to subscriber.
+func (s *stream) subscribe(callback func(batchedData []models.Record) error, batchSize int) *stream {
+	s.subscribers = append(s.subscribers, &subscriber{
+		callback:  callback,
+		batchSize: batchSize,
+		channel:   make(chan models.Record),
+	})
+
+	return s
+}
+
+// broadcast() will start listening to emitter for any pushed data.
+// This process is blocking, so most times you would want to call this inside a goroutine.
+func (s *stream) broadcast() error {
+	var wg sync.WaitGroup
+	for _, l := range s.subscribers {
+		wg.Add(1)
+		go func(l *subscriber) {
+			batch := newBatch(l.batchSize)
+			// listen to channel and emit data to subscriber callback if batch is full
+			for d := range l.channel {
+				batch.add(d)
+				if batch.isFull() {
+					if err := l.callback(batch.flush()); err != nil {
+						s.closeWithError(err)
+					}
+				}
+			}
+
+			// emit leftover data in the batch if any after channel is closed
+			if !batch.isEmpty() {
+				if err := l.callback(batch.flush()); err != nil {
+					s.closeWithError(err)
+				}
+			}
+
+			wg.Done()
+		}(l)
+	}
+
+	wg.Wait()
+
+	// there could be error when emitting
+	if s.err != nil {
+		return s.err
+	}
+
+	return nil
```

Comment:
Can't we do this?
```suggestion
	return s.err
```

Reply:
yeah totally can do that, there were a lot of modification before, missed that changes.